docs: document `tedge diag collect` user-guide (#3663)
thin-edge/thin-edge.io: docs/src/operate/troubleshooting/tedge-diag.md

| 01_tedge.sh        | Collects logs from `tedge-mapper` and `tedge-agent`              |
| 02_os.sh           | Collects system information                                      |
| 03_mqtt.sh         | Collects output of `tedge mqtt sub`                              |
| 04_workflow.sh     | Copies workflow logs                                             |
| 05_entities.sh     | Collects entity data                                             |
| 06_internal.sh     | Copies internal backup files                                     |
| 07_mosquitto.sh    | Collects mosquitto logs (skipped when using the built-in bridge) |
| template.sh.ignore | A template for creating custom diagnostic plugin                 |

### Writing your own diagnostic plugin

The easiest way to create a custom plugin is to start from the [template script](https://github.com/thin-edge/thin-edge.io/blob/main/configuration/contrib/diag-plugins/template.sh.ignore).

First, rename it to remove the `.ignore` extension and copy it to your desired location:

```sh
mkdir -p /etc/tedge/diag-plugins
cp /usr/share/tedge/diag-plugins/template.sh.ignore /etc/tedge/diag-plugins/100_my-plugin.sh
```

> Open the file in a text editor and modify the `collect()` function.

Reviewer: I would not state the obvious as "Open the file in a text editor" but the goal.

```suggestion
Then modify the `collect()` function to gather all relevant diagnostic data for your service, by printing status data and collecting log files.
```

Author: Applied in 8b134ec29235ed4d94ffe953e0b1ad2c0685221b

---
title: Collecting diagnostic logs
tags: [Operate]
sidebar_position: 1
description: How to collect logs and system diagnostics using plugin-based automation
---

> %%te%% helps you collect all diagnostic information about your device and services with a single command running user-defined diagnostic plugins.

Reviewer 2: ```suggestion
%%te%% diagnostic collector helps you collect all diagnostic information about your device and services with a single command which can be further extended with user-defined diagnostic plugins.
```
As this tool is not just about user-defined plugins but also the inbuilt plugins.

Author: Applied all your suggestions here 81c06445287266e448203d1b881a9ee1fedcb464.

The collected diagnostics are bundled into a single compressed archive (`.tar.gz`) for easy analysis or sharing with support.

This guide explains how to use the diagnostic collection command and how to add or customize diagnostic plugins.

## Quick Start

To collect a snapshot of diagnostic data, run:

```sh
tedge diag collect
```

This command performs the following steps:
* Discovers all available diagnostic plugins (default location: `/usr/share/tedge/diag-plugins`)

> * Executes each plugin in sequence

Reviewer 2: ```suggestion
* Executes each plugin in alphabetical order
```

pre-defined diagnostic plugins are installed in `/usr/share/tedge/diag-plugins`.

In addition, multiple plugin directories are supported.
You can specify multiple `--plugin-path` options:

```sh
tedge diag collect --plugin-path /usr/share/tedge/diag-plugins --plugin-path /your/own/dir/path
```

Alternatively, you can add a custom directory path permanently by `tedge config add` command.
In this case, you don't need to give specific path by command-line.

```sh
sudo tedge config add diag.plugin_paths "your/own/dir/path"
tedge diag collect
```

### Predefined plugins

The [predefined plugins](https://github.com/thin-edge/thin-edge.io/tree/main/configuration/contrib/diag-plugins) include:

| Plugin Name        | Description                                                                             |
| ------------------ | --------------------------------------------------------------------------------------- |
| 01_tedge.sh        | Collects logs from `tedge-mapper` and `tedge-agent`, along with `tedge config` settings |
| 02_os.sh           | Collects system information                                                             |

> | 03_mqtt.sh         | Collects output of `tedge mqtt sub`                                                     |

Reviewer 2: ```suggestion
| 03_mqtt.sh         | Collects all messages on the broker with `tedge mqtt sub #`                                                     |
```

---
title: Diagnostic Collection
tags: [Operate]
sidebar_position: 1
description: How to collect logs and system diagnostics using plugin-based automation
---

%%te%% helps you collect useful diagnostic information about your device and services by executing diagnostic plugins.
These diagnostics are bundled into a single compressed archive (`.tar.gz`) for easy analysis or sharing with support.

This guide explains how to use the diagnostic collection tool and how plugins work.

## Getting Started

To collect a snapshot of diagnostic data, run:

```sh
tedge diag collect
```

This command performs the following steps:
* Discovers all available diagnostic plugins (default location: `/usr/share/tedge/diag-plugins`)
* Executes each plugin in sequence
* Captures their logs and outputs

> * Bundles all collected data into a .`tar.gz` archive (e.g., `/tmp/tedge-diag-YYYY-MM-DD_HH-MM-SS.tar.gz`)

Reviewer: ```suggestion
* Bundles all collected data into a `tar.gz` archive (e.g., `/tmp/tedge-diag-YYYY-MM-DD_HH-MM-SS.tar.gz`)
```

All directory paths and the package name can be customized using command-line arguments or the `tedge config` command.

To see all available options, run:

```sh
tedge diag collect --help
```

> ## Diagnostic Plugins

Reviewer: This section gives the required content but in an order that can be improved.

- What is a diagnostic plugin
- Where to find them
- What is provided out-of-the-box
- What has to be done to add diagnostic plugins
- What is the contract between the runner and the plugins

Reviewer: Nicely done. I would just move the "Disabling plugins" section after the list of "Predefined plugins". The need to disable plugins comes only once aware of what is provided out-of-the-box.

> Diagnostic plugins are executable scripts or binaries invoked by `tedge diag collect` (referred to as the "runner") to gather diagnostic data.

Reviewer: ```suggestion
A diagnostic plugin is a command that collects a bunch of debug and diagnostic data, status and log files related to a process or service.
```

Each plugin is called with the collect subcommand, along with two options:
* `--output-dir`: where the plugin should store generated files
* `--config-dir:` the directory containing the configuration (`tedge.toml`)

Some diagnostic plugins are included with the %%te%% package and installed by default in `/usr/share/tedge/diag-plugins`.

> You can also create your own plugins and store them in any directory.

Reviewer: How to tell the runner where are my plugins? Can I have plugins in  several directories?

Each plugin is expected to finish within a limited time. If it runs too long, the runner will terminate it.

The plugin may print to `stdout` and `stderr`, which will be captured into an `output.log` file by the runner.
Additionally, it can write its own files into the specified output directory.

> ### Plugin Exit Codes

Reviewer: I would consider to skip this and let the user read the reference docs for these details.
